src/NuGet.Core/NuGet.Build.Tasks.Console/DependencyGraphSpecGenerator.cs

```diff
@@ -0,0 +1,802 @@
+
+            var frameworkDependencies = new List<FrameworkDependency>(frameworkReferenceItems.Count);
+
+            foreach (var frameworkReferenceItem in frameworkReferenceItems)
+            {
+                var privateAssets = MSBuildStringUtility.Split(frameworkReferenceItem.GetMetadataValue("PrivateAssets"));
+
+                frameworkDependencies.Add(new FrameworkDependency(frameworkReferenceItem.EvaluatedInclude, FrameworkDependencyFlagsUtils.GetFlags(privateAssets)));
+            }
+
+            return frameworkDependencies;
+        }
+
+        private static LibraryIncludeFlags GetLibraryIncludeFlags(string value, LibraryIncludeFlags defaultValue)
+        {
+            if (string.IsNullOrWhiteSpace(value))
+            {
+                return defaultValue;
+            }
+
+            string[] parts = MSBuildStringUtility.Split(value);
+
+            return parts.Length > 0 ? LibraryIncludeFlagUtils.GetFlags(parts) : defaultValue;
+        }
+
+        private static List<string> GetOriginalTargetFrameworks(IReadOnlyCollection<ProjectInstance> projectInnerNodes)
+        {
+            var targetFrameworks = new List<string>(projectInnerNodes.Count);
+
+            foreach (var projectInnerNode in projectInnerNodes)
+            {
+                targetFrameworks.Add(GetTargetFrameworkName(projectInnerNode));
+            }
+
+            return targetFrameworks;
+        }
+
+        private static IEnumerable<DownloadDependency> GetPackageDownloads(ProjectInstance projectInstance)
+        {
+            foreach (var projectItemInstance in projectInstance.GetItems("PackageDownload").Distinct(ProjectItemInstanceEvaluatedIncludeComparer.Instance))
+            {
+                string id = projectItemInstance.EvaluatedInclude;
+
+                foreach (var version in MSBuildStringUtility.Split(projectItemInstance.GetMetadataValue("Version")))
+                {
+                    VersionRange versionRange = !string.IsNullOrWhiteSpace(version) ? VersionRange.Parse(version) : VersionRange.All;
+
+                    if (!(versionRange.HasLowerAndUpperBounds && versionRange.MinVersion.Equals(versionRange.MaxVersion)))
+                    {
+                        throw new ArgumentException(string.Format(CultureInfo.CurrentCulture, "'{0}' is not an exact version like '[1.0.0]'. Only exact versions are allowed with PackageDownload.", versionRange.OriginalString));
+                    }
+
+                    yield return new DownloadDependency(id, versionRange);
+                }
+            }
+        }
+
+        private static List<LibraryDependency> GetPackageReferences(ProjectInstance projectInstance)
+        {
+            var packageReferenceItems = projectInstance.GetItems("PackageReference").Distinct(ProjectItemInstanceEvaluatedIncludeComparer.Instance).ToList();
```

Comment:
Are two PackageReferences same Id but different metadata equal?

Reply:
Is it expected that the target CollectPackageReferences was executed before this call?